hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/MisReplicationHandler.java

```diff
@@ -0,0 +1,204 @@
+
+  private Map<DatanodeDetails, SCMCommand<?>> getReplicateCommands(
+          ContainerInfo containerInfo,
+          Set<ContainerReplica> replicasToBeReplicated,
+          List<DatanodeDetails> targetDns) {
+    Map<DatanodeDetails, SCMCommand<?>> commandMap = Maps.newHashMap();
+    int datanodeIdx = 0;
+    for (ContainerReplica replica : replicasToBeReplicated) {
+      if (datanodeIdx == targetDns.size()) {
+        break;
+      }
+      commandMap.put(targetDns.get(datanodeIdx),
+              getReplicateCommand(containerInfo, replica));
+      datanodeIdx += 1;
+    }
+    return commandMap;
+
+  }
+  @Override
+  public Map<DatanodeDetails, SCMCommand<?>> processAndCreateCommands(
+      Set<ContainerReplica> replicas, List<ContainerReplicaOp> pendingOps,
+      ContainerHealthResult result, int remainingMaintenanceRedundancy)
+      throws IOException {
+    ContainerInfo container = result.getContainerInfo();
+    ContainerReplicaCount replicaCount = getContainerReplicaCount(container,
+            replicas, pendingOps, remainingMaintenanceRedundancy);
+
+    if (!replicaCount.isSufficientlyReplicated() ||
+            replicaCount.isOverReplicated()) {
+      LOG.info("Container {} state should be neither under replicated " +
+              "nor over replicated before resolving misreplication." +
+              "Container UnderReplication status: {}," +
+              "Container OverReplication status: {}",
+              container.getContainerID(),
+              !replicaCount.isSufficientlyReplicated(),
+              replicaCount.isOverReplicated());
+      return Collections.emptyMap();
+    }
+    final Set<DatanodeDetails> deletionInFlight = new HashSet<>();
+    final Set<DatanodeDetails> addInFlight = new HashSet<>();
+
+    for (ContainerReplicaOp op : pendingOps) {
+      if (op.getOpType() == ContainerReplicaOp.PendingOpType.DELETE) {
+        deletionInFlight.add(op.getTarget());
+      } else if (op.getOpType() == ContainerReplicaOp.PendingOpType.ADD) {
+        addInFlight.add(op.getTarget());
+      }
+    }
+
+    List<DatanodeDetails> usedDns = replicas.stream()
+            .map(ContainerReplica::getDatanodeDetails)
+            .collect(Collectors.toList());
+    if (containerPlacement.validateContainerPlacement(usedDns,
+            usedDns.size()).isPolicySatisfied()) {
+      LOG.info("Container {} is currently not misreplicated",
+              container.getContainerID());
+      return Collections.emptyMap();
+    }
+    List<DatanodeDetails> dnsWithPendingOps =
+            Stream.concat(usedDns.stream(), addInFlight.stream())
+            .filter(dn -> !deletionInFlight.contains(dn))
+            .distinct()
+            .collect(Collectors.toList());
+    if (containerPlacement.validateContainerPlacement(dnsWithPendingOps,
+            usedDns.size()).isPolicySatisfied()) {
+      LOG.info("Container {} is currently misreplicated currently" +
```

Comment:
" currently misreplicated currently" -> "currently misreplicated"

Reply:
Removed it